## Отчет по тестовому заданию

Автотест был реализован на Python 3.6 и Robot Framework 3.0.2


### Состав решения

library - папка с библиотекой на python
config - папка с variables.py файлом переменных автотеста
aut - папка с тестирумым приложением
tests - папка с тестом
report - папка с отчетами теста

### Требования и зависимости

Для запуска тестируемого приложения выполнить инструкции ./aut/README.md

Для запуска теста требуются:

        Python 3.6

    Установить зависимости при промощи команды:

        pip install -r requirements.txt

### Запуск автотеста

Запустите команду:

    robot -d report --variablefile config/variables.py tests

### Описание

Автотест реализован согласно техническому заданию.

При выполнении команды docker build файл Clients.db загружается в docker-образ и
база данных на локальном диске в ходе выполнения теста не обновляется. Автотест
определяет такие ошибки.

Перечень ошибок с описанием:

    1) "The client id Х does not have services." - сообщение выводится на Шаге
    "Wait appear new service for client" по истечении времени ожидания и пустом
    списке сервисов клиента.

    2) "Expected balance of client to be {expected balance} but was {current balance}" -
       сообщение выводится на Шаге "Check balance reduced to service cost",
       если текущий баланс из базы данных не соответствует ожидаемому балансу.
       Ожидаемый баланс вычисляется вычитанием из начального баланса
       стоимости подключенной услуги.

    3) "No unused services available" - сообщение выводится если для выбранного
       клиента с положительным балансом подключены все возможные услуги.

Рекомендации для устранения ошибок и успешного прохождения теста:

    Ошибки 1 и 2 - возникают из-за отличия баз данных на локальном диске и в
    docker-контейнере.

    Ошибка 1 - если отсутствует клиент с положительным балансом, автотест добавляет
    нового клиента в базу данных на локальном диске, но изменение не отражается в
    docker-контейнере.

    Ошибка 2 - добавление нового сервиса в базу данных на локальном диске не отражается
    в docker-контейнере

       Ошибки 1 и 2 исправляются включением в файл docker-compose.yml:

           web:
             ...
             volumes:
                - ./web:/app/

   Ошибка 3 - возникает если все услуги уже подключены для клиента.

   Ошибка 3 исправляется изменением теста. В ходе каждого теста в базу данных
   добавляется новый клиент.





